### ISSUE TITLE
FMC ACP rule hit count - feature request

### ISSUE THREAD

**Issue by karmicgeezer** [daxm/fmcapi#39]
Are there any plans to use the 6.4 features to be able to get ACP rule hit counts?

**Comment by daxm** [comment]
Our goal is to eventually support access to ALL FMC API objects.  That said, we do them as we can.  Are you interested/willing to help the project out and work on that?

**Comment by karmicgeezer** [comment]
I can certainly try, I would love to be able to contribute as a great project!

**Comment by daxm** [comment]
Any progress on this feature?  If not, I'll give it a try.

**Comment by daxm** [comment]
hitcount and prefilter_policy GET features are now in Master branch.  However, hitcount also can do PUT or DELETE but I haven't written those features into fmcapi yet.  Test it out and see what you think!

**Comment by NetDevAutomate** [comment]
Do you have an example/unit test for hit count? There was nothing in the repo and the code base has had a pretty major refactor.

**Comment by daxm** [comment]
I’m not in front of my computer right now but I’ll describe that later 

It is far better to grasp the Universe as it really is than to persist in delusion, however satisfying and reassuring.  — Carl Sagan

**Comment by daxm** [comment]
Yes, we did a major refactor to better organize the code.  Hopefully (other than trying to find things) this reorg didn't break your workflow!

I must have lost some code or "thought" HitCount was done.  I just spent the day getting the GET features to work.  There is still an issue that I can't seem to resolve (you can't add an AccessRule to the HitCount's filter via the rule's name... it provides/expects the wrong value!)
Anyway, here is the sample code:  https://github.com/daxm/fmcapi/blob/master/unit_tests/hit_counts.py

Note: These changes aren't published to PyPi yet so you'll have to clone the repo to try them out.

**Comment by daxm** [comment]
Note the """ quoted note near the end of the code.  I haven't figured out why but the HitCount "ID" for the AccessRule doesn't match what the AccessRule's ID is.  Try using acp_name instead.

**Comment by daxm** [comment]
Check out the debug output for what is returned from GET'ting the AccessRule vs what the HitCounts sees as the id for that particular AccessRule.

**Comment by NetDevAutomate** [comment]
Same problem when using name when using:

acprule1 = fmcapi.AccessRules(fmc=fmc, acp_name=device1.accessPolicy['name'])

Log file:
2019/11/04-15:53:36 - INFO:apiclasstemplate.py:135 - GET success. Object with name: "<device name>" and id: "<device id>" fetched from FMC.
2019/11/04-15:53:36 - INFO:apiclasstemplate.py:135 - GET success. Object with name: "<policy name>" and id: "<policy id>" fetched from FMC.
2019/11/04-16:09:27 - INFO:apiclasstemplate.py:135 - GET success. Object with name: "<device name>" and id: "<device id>" fetched from FMC.
2019/11/04-16:09:27 - WARNING:fmc.py:245 - Either:
	1. Payload too large.  FMC can only handle a payload of 2048000 bytes.
	2.The payload contains an unprocessable or unreadable entity such as a invalid attribut name or incorrect JSON syntax
2019/11/04-16:09:27 - ERROR:fmc.py:253 - Error in POST operation --> 422 Client Error: Unprocessable Entity for url: https://<IP address>/api/fmc_config/v1/domain/<domain>/policy/accesspolicies/<ACP>/operational/hitcounts?filter=%22deviceId:<device id>%22&expanded=true&limit=500?filter=%22deviceId:<device id>%22&expanded=true&limit=500
2019/11/04-16:09:27 - ERROR:fmc.py:254 - json_response -->	{'error': {'category': 'VALIDATION', 'messages': [{'description': 'The limit in query is not a number.', 'code': 'invalidLimit', 'location': 'limit'}], 'severity': 'ERROR'}}
2019/11/04-16:09:27 - INFO:fmc.py:154 - Auto deploy changes set to False.  Use the Deploy button in FMC to push changes to FTDs.

**Comment by daxm** [comment]
Yeah.  I corrected my last message.

It is far better to grasp the Universe as it really is than to persist in delusion, however satisfying and reassuring.  — Carl Sagan

**Comment by NetDevAutomate** [comment]
When using acprule1 = fmcapi.AccessRules(fmc=fmc, acp_name=device1.accessPolicy["name"])

The log shows the FTD device and ACP rule IDs, using acp_id=device1.accessPolicy["id"]) only the FTD device ID is found.

Neither option get's as far as showing HitCount logs.

**Comment by daxm** [comment]
I'm going to have to look into this some more (obviously).  I know I can 
get it to work if I already know the "correct" ID that HitCounts is 
looking for.  That said, I don't know how to get that ID without first 
pulling all the HitCounts.  I need to find someone who knows what is 
going on here.

**Comment by daxm** [comment]
Thank you.  That said, you aren't off the hook.  Maybe you can help me 
figure out what ID the HitCount is using when it references the 
AccessRule.  :-)

**Comment by NetDevAutomate** [comment]
I knocked up some rough python and used:

```
api_path = f"/api/fmc_config/v1/domain/{domain}/policy/accesspolicies/{acp}/operational/hitcounts?filter=%22deviceId%3A{deviceID}%3BfetchZeroHitCount%3A{ZeroCount}%22&expanded={expanded}&limit={LIMIT}"
```

acp = ACP ID

The code isn't pretty but did give me hit counts for each ACP entry.

When I set ZeroCount to true oddly enough got nothing, setting it to false I got everything including ZeroCount which is actually the only thing I was interested in.

**Comment by MysticRyuujin** [comment]
So I just tried to use this, and it works...but there are a few things wrong with it (and one thing wrong with the actual API itself...)

First what's wrong with the API itself? Using version 6.6.1 of the FMC, the hitcounts endpoint will not return more than 1,000 objects, even if the limit is set to be greater than 1,000. Additionally, it does not page. It will always say that the limit was set to 1,000 in the reply and will only return 1,000 objects. Basically if you send limit=2000, the response link to self will say limit=1000

Now, back to this module...

There are 2 things not working:
1. There is no offset or index number available. This is important because of the issue above, if I get back 1,000 objects, there's a very good chance that there are more objects that I need to go get, but I can't because I can't set an offset/index. The module doesn't know there are more objects to get because the paging is broken :smile:  so there needs to be a way to set an offset in the module...

2. You cannot get an HA pair device using its name. I don't know if this is a bigger problem than just the hitcounts endpoint, but using the device_name only works for non-HA pairs. You must use the deviceId of the HA pair as the containerId

**Comment by Niltak** [comment]
Offset was added in 'Added searches, updated natrules, and fixed minor issues with hitcounts and apiclasstemplate #127'
